### PR TITLE
Chapter/chapter2 working with data

### PR DESCRIPTION
**Pull request by hothanhtuan1209**
In this pull request, I read docstring,  compare harmful and idiomatic code for chapter 2 in ebook write_idiomatic_python

### CODE REVIEW

File: woking_with_data/classes/define_(__str__)_in_class.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+#         self.first_value = first_value
+#         self.second_value = second_value
+
+# point = Point(1, 2)
+# print(point)
+# Prints '<__main__.Point object at 0x91ebd0>'
+
+
+# Idiomatic solution
+class Point():
```
Comment (cuongdoan-agilityio) [inline woking_with_data/classes/define_(__str__)_in_class.py]: Missing docstring.

File: woking_with_data/classes/don't_use_classes_as_encapsulation.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+def count_occurrences(string, key):
+    """
+    Counts the number of occurrences of a character or substring in the input string.
+
+    Args:
+        string (str): The string to count occurrences in.
+        key (str): The character or substring to count.
+
+    return:
```
Comment (cuongdoan-agilityio) [inline woking_with_data/classes/don't_use_classes_as_encapsulation.py]: I see you use two different ways of writing (Line 61), please agree on one way of writing, and check for all other files.

File: woking_with_data/classes/don't_use_classes_as_encapsulation.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+def is_palindrome(string):
+    """
+    Checks if a string is a palindrome (reads the same forwards and backwards).
+
+    Args:
+        string (str): The string to check.
+
+    Returns:
+        bool: True if 'string' is a palindrome, False otherwise.
```
Comment (cuongdoan-agilityio) [inline woking_with_data/classes/don't_use_classes_as_encapsulation.py]: What is the `pool` ?

File: woking_with_data/classes/use_(__repr__).py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+#         return 'Bar is {}, Baz is {}'.format(self.bar, self.baz)
+
+# def log_to_console(instance):
+#     print(instance)
+
+# log_to_console([Foo(), Foo(cache={'x': 'y'})])
+
+
+# Idiomatic solution
+class Foo():
```
Comment (cuongdoan-agilityio) [inline woking_with_data/classes/use_(__repr__).py]: Missing docstring, Please add docstring for all classes and functions.

File: woking_with_data/classes/use_underscores.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        self.__id = 8
+        self.value = self.__get_value() # Our 'private copy'
+    
+    def get_value(self):
+        pass
+    
+    def should_destroy_earth(self):
+        return self.__id == 42
+    
+    # Here, we're storing a 'private copy' of get_value,
```
Comment (cuongdoan-agilityio) [inline woking_with_data/classes/use_underscores.py]: Should explain briefly.

File: woking_with_data/strings/use_ord_ascii.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+# }
+
+# for e in some_string:
+#     hash_value += character_hash[e]
+
+
+# Idiomatic solution
+hash_value = 0
+
+for e in some_string:
```
Comment (cuongdoan-agilityio) [inline woking_with_data/strings/use_ord_ascii.py]: I haven't seen you define `some_string`

Comment (hothanhtuan1209) [reply woking_with_data/strings/use_ord_ascii.py]: Done

File: woking_with_data/classes/define_(__str__)_in_class.py (hunk 1 above)
Comment (hothanhtuan1209) [reply woking_with_data/classes/define_(__str__)_in_class.py]: Done

File: woking_with_data/classes/use_(__repr__).py (hunk 4 above)
Comment (hothanhtuan1209) [reply woking_with_data/classes/use_(__repr__).py]: Done

File: woking_with_data/classes/don't_use_classes_as_encapsulation.py (hunk 2 above)
Comment (hothanhtuan1209) [reply woking_with_data/classes/don't_use_classes_as_encapsulation.py]: Done

File: woking_with_data/classes/use_underscores.py (hunk 5 above)
Comment (hothanhtuan1209) [reply woking_with_data/classes/use_underscores.py]: Done

**Pull request merged**